linked_list/linked_list.py

```diff
@@ -2,96 +2,173 @@
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        current = self.head
+
+        while current:
+            if current.value == value:
+                return True
+            current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def length(self):
-        pass
+        count = 0
+        current = self.head
+
+        while current != None:
+            count += 1
+            current = current.next
+        return count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def get_at_index(self, index):
-        pass
+        if self.length() < index:
+            return None
+
+        current = self.head
+        current_index = 0
+        while current != None:
+            if current_index == index:
+                return current.value
+            current = current.next
+            current_index += 1
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        current = self.head
+
+        if current == None:
+            return None
+
+        while current.next != None:
+            current = current.next
+        return current.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
+
     def add_last(self, value):
-        pass
+        new_node = Node(value)
+        current = self.head
+        if not self.head:
+            self.add_first(value)
+
+        else:
+
+            while current.next != None:
+                current = current.next
+
+            current.next = new_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+
     def find_max(self):
-        pass
+        current = self.head
+        
+        if current == None:
+            return None
+        max = self.head.value
+        
+        while current != None:
+            if current.value > max:
+                max = current.value
+            current = current.next
+
+        return max
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
+
     def delete(self, value):

```

Comment:
👍 , but the time complexity is O(n) as you have to *find* the node to delete it.